### ISSUE TITLE
ENTRYPOINT with heredoc created file does not pass additional arguments to it

### ISSUE THREAD

**Issue by udayh** [docker/cli#4127]
### Description

Hi - I installed Docker version 23.0.1, build a5ee5b1 using the instructions on the website, in order to take advantage of buildx and heredoc features. I have seen two issues with it so far

1. I created a heredoc using:
COPY --chown=${U}:${U} --chmod=755 <<EOT some_filename
...
content
...
EOT

I then named it as an ENTRYPOINT (exec-style). The script runs when I run the container, but any additional arguments I pass on the 'run' command line do not get passed to it. If I move the content to an actual file that gets COPY'd it works as expected

2. I had a multiline invocation of a command in the heredoc; I expected to use single backslashes to merge the multiple lines. However, I had to use double-backslashes to be able to corral all the lines

### Reproduce

See description

### Expected behavior

See description

### docker version

```bash
Client: Docker Engine - Community
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        a5ee5b1
 Built:             Thu Feb  9 19:47:01 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       bc3805a
  Built:            Thu Feb  9 19:47:01 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.16.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 5
 Server Version: 23.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2456e983eb9e37e47538f59ea18f2043c9a73640
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.0-custom
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 23.35GiB
 Name: amd-Celadon-RN
 ID: ALXL:LQOH:7FX7:O2WH:4PG3:FHKP:UGZJ:WTAT:LYCT:K7SZ:OMT7:R7SF
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by bsousaa** [comment]
Can you provide a full example with the exact steps to reproduce? As this is related with build we are moving to the Buildx repo.

**Comment by thaJeztah** [comment]
/cc @crazy-max @dvdksn perhaps this is something we should try to include somewhere in our docs (if we don't have anything about this).

TL;DR a "shell form" entrypoint implicitly means `sh -c '<entrypoint>`. And arguments passed to the container when running thus means "pass arguments to `sh -c` (_NOT_ to `<entrypoint>`)

**Comment by udayh** [comment]
I think the issue was using $1 rather than \$1, thus everything in the heredoc has to be escaped which is unexpected. Will report against buildx

**Comment by thaJeztah** [comment]
To prevent heredoc from expanding / substituting, you need to put the `WORD` in quotes (same applies if you'd use them in Bash)

So, for example;

```dockerfile
COPY  --chmod=755 <<'EOT' /entry.sh
#!/bin/sh
echo oh hi $1
EOT
```

(notice the quotes around `'EOT'`)

**Comment by thaJeztah** [comment]
Also see the "Example 19-7. Parameter substitution turned off" section here: https://tldp.org/LDP/abs/html/here-docs.html
